File: ios/GradeThread/Marketplaces/Publish/EbayCondition.swift
import Foundation

/// eBay item-condition enum used by the publish composer. Raw values are
/// the exact strings eBay's Inventory API expects (and what the edge stores
/// in `listings.ebay_condition`). Mirrors the web composer's dropdown.
enum EbayCondition: String, CaseIterable, Identifiable {
    case new = "NEW"
    case likeNew = "LIKE_NEW"
    case usedExcellent = "USED_EXCELLENT"
    case usedVeryGood = "USED_VERY_GOOD"
    case usedGood = "USED_GOOD"
    case usedAcceptable = "USED_ACCEPTABLE"
    case forParts = "FOR_PARTS_OR_NOT_WORKING"

    var id: String { rawValue }

    var label: String {
        switch self {
        case .new:            return "New with tags"
        case .likeNew:        return "New without tags"
        case .usedExcellent:  return "Pre-owned – Excellent"
        case .usedVeryGood:   return "Pre-owned – Very Good"
        case .usedGood:       return "Pre-owned – Good"
        case .usedAcceptable: return "Pre-owned – Acceptable"
        case .forParts:       return "For parts / not working"
        }
    }

    /// Resolves a server condition string to a case, defaulting to
    /// "Pre-owned – Excellent" (the most common resale condition) when the
    /// value is missing or unrecognized. Use this only where a concrete
    /// condition must be chosen (e.g. the publish composer's initial picker);
    /// for display, prefer ``displayLabel(for:)`` so an unrecognized value
    /// isn't misrepresented as "Excellent".
    static func resolve(_ raw: String?) -> EbayCondition {
        guard let raw, let match = EbayCondition(rawValue: raw) else {
            return .usedExcellent
        }
        return match
    }

    /// A human label for a stored condition string that invents NO fallback
    /// (US-1268): a recognized value maps to its label; an unrecognized but
    /// non-empty value is shown verbatim so a row never misrepresents what's
    /// actually stored. Returns nil for a missing/blank value (nothing to show).
    static func displayLabel(for raw: String?) -> String? {
        guard let trimmed = raw?.trimmingCharacters(in: .whitespacesAndNewlines),
              !trimmed.isEmpty else {
            return nil
        }
        return EbayCondition(rawValue: trimmed)?.label ?? trimmed
    }
}

File: android/app/src/main/java/com/gradethread/app/marketplaces/publish/EbayCondition.kt
package com.gradethread.app.marketplaces.publish

/**
 * US-1352: eBay item condition for the publish composer. [wire] is the exact
 * string eBay's Inventory API expects and what the edge stores in
 * `listings.ebay_condition` — mirrors the web composer's dropdown and iOS
 * `EbayCondition`.
 */
enum class EbayCondition(val wire: String, val label: String) {
    NEW("NEW", "New with tags"),

    /**
     * "New without tags" is NEW_OTHER (eBay id 1500), NOT LIKE_NEW (2750):
     * most clothing categories reject 2750, so labelling likeNew as "new
     * without tags" sends an invalid condition (eBay error 25021).
     */
    NEW_OTHER("NEW_OTHER", "New without tags"),
    NEW_WITH_DEFECTS("NEW_WITH_DEFECTS", "New with defects"),
    LIKE_NEW("LIKE_NEW", "Like new"),
    USED_EXCELLENT("USED_EXCELLENT", "Pre-owned – Excellent"),
    USED_VERY_GOOD("USED_VERY_GOOD", "Pre-owned – Very Good"),
    USED_GOOD("USED_GOOD", "Pre-owned – Good"),
    USED_ACCEPTABLE("USED_ACCEPTABLE", "Pre-owned – Acceptable"),
    FOR_PARTS("FOR_PARTS_OR_NOT_WORKING", "For parts / not working"),
    ;

    companion object {
        /**
         * Resolves a stored condition to a case, defaulting to the most common
         * resale condition. Use ONLY where a concrete choice must be made (the
         * composer's initial picker) — for display prefer [displayLabel], which
         * invents nothing.
         */
        fun resolve(raw: String?): EbayCondition =
            entries.firstOrNull { it.wire == raw?.trim() } ?: USED_EXCELLENT

        /**
         * A label for a stored condition with NO invented fallback: a known
         * value maps to its label, an unknown but non-blank value is shown
         * verbatim so a row never misrepresents what is stored, and a blank one
         * returns null (there is nothing to show).
         */
        fun displayLabel(raw: String?): String? {
            val trimmed = raw?.trim()?.takeIf { it.isNotEmpty() } ?: return null
            return entries.firstOrNull { it.wire == trimmed }?.label ?: trimmed
        }
    }
}
